bindings/Lua/judger.c

```diff
@@ -0,0 +1,219 @@
+
+    lua_createtable(L, 0, conf_item_size);
+    for (; i < conf_item_size; ++i) {
+        lua_pushstring(L, conf_item[i].name);
+        lua_pushnumber(L, conf_item[i].value);
+        lua_rawset(L, -3);
+    }
+    assert(lua_gettop(L) == 1 && lua_type(L, -1) == LUA_TTABLE); //One table left
+
+    return 1;
+}
+
+static void destory_config(struct config* config) {
+    int i = 0;
+
+    free(config->exe_path);
+    free(config->input_path);
+    free(config->output_path);
+    free(config->error_path);
+    free(config->log_path);
+    free(config->seccomp_rule_name);
+
+    for (; i < 256; ++i) {
+        free(config->args[i]);
+        free(config->env[i]);
+    }
+}
+
+static int lua_checkstringtable(lua_State* L, int index, char** dst, int max_size, int* dst_size) {
+    int table_index = lua_gettop(L);
+    int table_size = 0;
+
+    lua_pushnil(L);
+    while (0 != lua_next(L, table_index)) {
+        if (lua_type(L, -1) != LUA_TSTRING)
+            return 0;
+
+        if (table_size == max_size)
+            return 0;
+
+        dst[table_size] = _strdup(lua_tostring(L, -1));
+        table_size++;
+
+        lua_pop(L, 1);
+    }
+
+    if (dst_size != NULL)
+        *dst_size = table_size;
+
+    return 1;
+}
+
+static int global_judge(lua_State* L) {
+    int i = 0, table_size = 0;
+    struct config config;
+    struct result result;
+    void* arg_refs[arg_mask_size];
+
+    arg_refs[0]  = &config.max_cpu_time;
+    arg_refs[1]  = &config.max_real_time;
+    arg_refs[2]  = &config.max_memory;
+    arg_refs[3]  = &config.max_process_number;
+    arg_refs[4]  = &config.max_output_size;
+    arg_refs[5]  = &config.exe_path;
+    arg_refs[6]  = &config.input_path;
+    arg_refs[7]  = &config.output_path;
+    arg_refs[8]  = &config.error_path;
+    arg_refs[9]  = &config.log_path;
+    arg_refs[10] = &config.seccomp_rule_name;
+    arg_refs[11] = &config.args;
+    arg_refs[12] = &config.env;
+    arg_refs[13] = &config.uid;
+    arg_refs[14] = &config.gid;
+    memset(&config, 0, sizeof(config));
+
+    luaL_checktype(L, 1, LUA_TTABLE);
+    for (; i < arg_mask_size; ++i) {
+        lua_pushstring(L, arg_mask[i].name);
+        lua_gettable(L, -2);
+        if (lua_type(L, -1) != arg_mask[i].t && !(arg_mask[i].nil_accept && lua_type(L, -1) == LUA_TNIL))
+            return destory_config(&config), luaL_error(L, "bad argument %s", arg_mask[i].name);
+
+
+        switch (lua_type(L, -1)) {
+        case LUA_TNUMBER:
+            *(int*)arg_refs[i] = lua_tointeger(L, -1);
+            break;
+        case LUA_TSTRING:
+            *(char**)arg_refs[i] = _strdup(lua_tostring(L, -1));
+            break;
+        case LUA_TTABLE:
+            if (lua_checkstringtable(L, -1, (char**)arg_refs[i], 256 - 1, &table_size) == 0) {
+                destory_config(&config);
+                return luaL_error(L, "bad argument %s (item must be a string)", arg_mask[i].name);
+            }
+            *((char**)arg_refs[i] + table_size) = NULL;
```

Comment:
Opps..I have not considered NIL at first.Fixed.
Seems it's better to have a `break` in the last case.